embedded_rtic/examples/sonars.rs

```diff
@@ -0,0 +1,440 @@
+                        }
+                        cx.local.times[0] = zero;
+                    }
+                }
+                SonarChannel::Left => {
+                    //info!("Event triggered by Sonar 2 (Left)");
+                    if cx.local.times[1] == zero {
+                        cx.local.times[1] = time;
+                    } else {
+                        let distance2 = time.checked_duration_since(cx.local.times[1]).unwrap().to_micros() / 56;
+                        cx.local.times[1] = zero;
+                        match cx.local.sender2.try_send(distance2 as u32) {
+                            Ok(_) => {}
+                            _ => {
+                                error!("Distance LEFT did not fit in to the buffer, scheduling is probably broken.");
+                                // This is not fatal so we simply continue with our life
+                            }
+                        }
+                    }
+                }
+                SonarChannel::Right => {
+                    //info!("Event triggered by Sonar 3 (Right)");
+                    if cx.local.times[2] == zero {
+                        cx.local.times[2] = time;
+                    } else {
+                        let distance3 = time.checked_duration_since(cx.local.times[2]).unwrap().to_micros() / 56;
+                        info!("distance 3 before sending: {:?}", distance3);
+                        cx.local.times[2] = zero;
+                        match cx.local.sender3.try_send(distance3 as u32) {
+                            Ok(_) => {}
+                            _ => {
+                                error!("Distance RIGHT did not fit in to the buffer, scheduling is probably broken.");
+                                // This is not fatal so we simply continue with our life
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        // Clear interrupt pending bits.
+        cx.shared.gpiote.lock(|gpiote| {
+            gpiote.reset_events();
+            gpiote.channel0().clear();
+            gpiote.channel1().clear();
+            gpiote.channel2().clear();
+        });
+    }
+
+    #[task(priority = 2, local = [trig], shared = [gpiote])]
+    /// Send a small pulse to the sonars.
+    ///
+    /// The sonars will then notify us in echo when the sound waves are
+    /// recieved.
+    async fn trigger1(cx: trigger1::Context) {
+        // Set high is always valid.
+        cx.local.trig.set_high().unwrap();
+
+        let now = Mono::now();
+        Mono::delay_until(now + 20.micros()).await;
+
+        // Set low is always valid.
+        cx.local.trig.set_low().unwrap();
+    }
+
+    #[task(priority = 2, local = [trig2], shared = [gpiote])]
+    /// Send a small pulse to the sonars.
+    ///
+    /// The sonars will then notify us in echo when the sound waves are
+    /// recieved.
+    async fn trigger2(cx: trigger2::Context) {
```

Comment:
it might be better to do something like


```rust

    #[task(priority = 2)]
    /// Send a small pulse to the sonars.
    ///
    /// The sonars will then notify us in echo when the sound waves are
    /// recieved.
    async fn trigger(cx: trigger2::Context,trigger_pin:&mut Pin<Output<PushPull>>) {
       ...
    }
    ```
    And then passing the specific pin from the specific trigger trampoline, reducing the complexity of the app and making it easier to read
    See rtic docs for more details
    https://rtic.rs/2/book/en/by-example/software_tasks.html